### ISSUE TITLE
ESP32: socket.recv() and socket.read() sometimes hang forever

### ISSUE THREAD

**Issue by chrisb2** [micropython/micropython#3754]
I have implemented Response.iter_content() in urequests so that I can deal with large responses without running out of memory. I based my code on the equivalent in [Python requests](https://github.com/requests/requests/blob/2de907ad778de270911acaffe93883f0e2729a4a/requests/models.py#L724). My new method is:
```
    def iter_content(self, chunk_size=1):
        def generate():
            while True:
                chunk = self.raw.read(chunk_size)
                if not chunk:
                    break
                yield chunk
            self._content_consumed = True

        if self._content_consumed:
            raise RuntimeError("response already consumed")
        elif chunk_size is not None and not isinstance(chunk_size, int):
            raise TypeError("chunk_size must be an int, it is instead a %s."
                % type(chunk_size))

        return generate()
```
I copied urequests and renamed to requests before adding iter_content() to requests, so that I could have both modules for testing.

My test program that executes this is:
```
import requests

for x in range(1,20):
    r = requests.get('http://jsonplaceholder.typicode.com/users')
    try:
        for chunk in r.iter_content(chunk_size=256):
            print(chunk.decode('UTF-8'))
    finally:
        r.close()
```
Running this with either recv() or read() in iter_content() I get the following results:
- Loboris Micropython for ESP32 - always hangs before 20 iterations is reached
- Micropython for ESP32 - always hangs before 20 iterations is reached
- Micropython for ESP8266 - never hangs
- Python3 (std requests on Raspbian on RPi) - never hangs

The hang always occurs when the request looks complete in terms of what is printed out. If I press cntl-C when a hang has occurred I get:
```
Traceback (most recent call last):
  File "<stdin>", line 7, in <module>
  File "<stdin>", line 4, in <module>
  File "requests.py", line 36, in generate
KeyboardInterrupt:
```
where line 36 is 
`chunk = self.raw.read(chunk_size)`

regards,
Chris

**Comment by dpgeorge** [comment]
Thanks for the report, and thanks for the info to reproduce the bug.  I could reproduce it myself and found that the cause was because lwIP in the ESP IDF does not have POSIX behaviour for recvfrom: it blocks forever if a peer closed the socket and additional reads are made on that socket.  This has now been fixed in 6681eb809a815a049ca069d522b9ad07548db535.  See the commit for details.

**Comment by chrisb2** [comment]
Confirmed this fixes the problem, thank you for amazing quick fix!
